README.md
# myca-cli

A command line for [myca](https://www.npmjs.com/package/myca) for creating my CA center, 
generating a self signed x509 certificate, issuing server certificate from node.js via openssl. 
Multiple center supported. RSA, EC(P-256, P-384) supported.

[![Version](https://img.shields.io/npm/v/myca-cli.svg)](https://www.npmjs.com/package/myca-cli)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)


## Installing

```bash
npm install -g myca-cli
```

## myca module
- [myca](https://www.npmjs.com/package/myca)


## Usage

- Initialize default center
  ```bash
  myca init
  ```

  will output:
  ```bash
  Default center created at path: "C:\Users\<user>\.myca"
  ```

- Initialize CA cert of default center
  ```bash
  myca initca --days=10950 --pass=mycapass \
    --cn="my root ca" --o="my company" --c=CN \
  ```
  will output:
  ```bash
  CA certificate created with:
    centerName: "default"
    crtFile: "C:\Users\<user>\.myca\ca.crt"
    privateKeyFile: "C:\Users\<user>\.myca\ca.key"
  ```

- Issue a RSA serve certificate
  ```bash
  myca issue --kind=server --days=730 --pass=fooo \
    --cn="waitingsong.com" --o="my company" --c=CN --caKeyPass=mycapass \
  ```

  will output:
  ```bash
  Issue a Certificate with:
    pubKey:
  -----BEGIN PUBLIC KEY-----
  MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxJunjvIoZ5bFQsA9D/1A
  MHt36viM7AJZFpQdmVuTLUZXEiTFU6gMdBarikHsXt0xRPcnGHiP1hgSsTIh2j1k
  3HiNinwfV/MePvy/8f/XWY+J3BbljQCPQmtUIZAnBebiVcvQrL1cP4l5xgJiv5/p
  EdRhCs92J/1MMDxhp41BzatBKwbQJ7UQtLnTdWXCs/qptTgaD6vh4a3snWHlfatg
  TsfzjmSmiXcEYGZM9z6tDrSjR9kBZoog+9DTh+FCdVaasL7QvYlWlOzsjSO2yvLX
  lYQJ9VJbBGxV0cOKbmPm46aMK6n5br/75CAm8cHyfgsE0MhxH2uxQW3leUy+3MHK
  ZwIDAQAB
  -----END PUBLIC KEY-----

    pass: "fooo"
    privateKeyFile: "C:\Users\<user>\.myca\server\01.key"
    privateUnsecureKeyFile: "C:\Users\<user>\.myca\server\01.key.unsecure"
    centerName: "default"
    caKeyFile: "C:\Users\<user>\.myca\ca.key"
    caCrtFile: "C:\Users\<user>\.myca\ca.crt"
    csrFile: "C:\Users\<user>\.myca\server\01.csr"
    crtFile: "C:\Users\<user>\.myca\server\01.crt"
  ```

- Initialize a center named ec
  ```bash
  // path can be omitted
  myca initcenter --name=ec --path="c:/users/<user>/.myca-ec"
  ```

  will output:
  ```bash
  center created with:
    centerName: "ec"
    path: "c:/users/<user>/.myca-ec"
  ```

- Create self-signed EC CA certificate under center ec (default P-256)
  ```bash
  myca initca --days=10950 --pass=mycapass \
    --cn="my root ca" --o="my company" --c=CN --centerName=ec --alg=ec \
  ```

  will output:
  ```bash
  CA certificate created with:
    centerName: "ec"
    crtFile: "c:\users\<user>\.myca-ec\ca.crt"
    privateKeyFile: "c:\users\<user>\.myca-ec\ca.key"
  ```

- Issue a ec server certificate by center ec CA cert
  ```bash
  myca issue --kind=server --days=730 --pass=fooo \
    --cn="foo.waitingsong.com" --o="my comany" --c=CN --caKeyPass=mycapass \
    --centerName=ec --alg=ec \
  ```

- Issue a serve certificate with Domain Name SANs
  ```bash
  myca issue --kind=server --days=730 --pass=fooo \
    --cn="waitingsong.com" --o="my company" --c=CN --caKeyPass=mycapass \
    --SAN="foo.waitingsong.com, bar.waitingsong.com" \
  ```

- Issue a serve certificate with IP SANs
  ```bash
  myca issue --kind=server --days=730 --pass=fooo \
    --cn="waitingsong.com" --o="my company" --c=CN --caKeyPass=mycapass \
    --ips=192.168.0.1 \

  myca issue --kind=server --days=730 --pass=fooo \
    --cn="waitingsong.com" --o="my company" --c=CN --caKeyPass=mycapass \
    --ips="192.168.0.1, 192.168.0.2" \
  ```

- Issue a RSA client p12/pfx certificate
  ```bash
  myca issue --kind=client --days=730 --pass=fooo \
    --cn="waitingsong.com" --o="my company" --c=CN --caKeyPass=mycapass \
  ```

  will output:
  ```bash
  Issue a Certificate with:
    pubKey:
  -----BEGIN PUBLIC KEY-----
  MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsu8wZDZ0a/HNtlJPqCjs
  9Isg795iUAJ+5OREb08hPthDN4/LOoLgepIyWbZ/A+0Gv8jHkbqlUvOJV5O5ggjR
  ezpK3jXln621nbjS3Fzs/uw4+40e4RX7fYIoE9sk94rP+od1ZMRjE8+e+qb34ubC
  WiXtsyR4EyaRen23IqLNlvxGlcg4xLczaCDA06zkva+wL7qvLYF2331X/rZ+dQgY
  xh6iWKO7C9qcliF23OOByYIKS8jqQ8ngwHIEogIqNBdt/QyEVN7CvF4M6abQnrrx
  9wnnmlaRX2WiybsA06wWl7+4BgKjeULehCVQOpMsS/3QV1dO79vn9hZWM/dAPlnF
  QwIDAQAB
  -----END PUBLIC KEY-----

    pass: "fooo"
    privateKeyFile: "C:\Users\<user>\.myca\client\0A.key"
    centerName: "default"
    caKeyFile: "C:\Users\<user>\.myca\ca.key"
    caCrtFile: "C:\Users\<user>\.myca\ca.crt"
    csrFile: "C:\Users\<user>\.myca\client\0A.csr"
    crtFile: "C:\Users\<user>\.myca\client\0A.crt"
    pfxFile: "C:\Users\<user>\.myca\client\0A.p12"
  ```


## License

[MIT](LICENSE)


## Languages

- [English](README.md)
- [中文](README.zh-CN.md)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.2.0"></a>
# [3.2.0](https://github.com/waitingsong/node-myca-cli/compare/v3.1.1...v3.2.0) (2019-01-23)


### Features

* genCmdHelp(command) command can be empty string ([b33e489](https://github.com/waitingsong/node-myca-cli/commit/b33e489))



<a name="3.1.1"></a>
## [3.1.1](https://github.com/waitingsong/node-myca-cli/compare/v3.1.0...v3.1.1) (2019-01-22)



<a name="3.1.0"></a>
# [3.1.0](https://github.com/waitingsong/node-myca-cli/compare/v3.0.0...v3.1.0) (2019-01-22)


### Features

* do isPathAccessible() first within isDirFileExists() ([9ddae98](https://github.com/waitingsong/node-myca-cli/commit/9ddae98))
* initcenter() path can be omitted ([3df744a](https://github.com/waitingsong/node-myca-cli/commit/3df744a))


test/10_index.test.ts
/// <reference types="mocha" />

import {
  basename,
  isPathAccessible,
  join,
  normalize,
  promisify,
  userHome,
} from '@waiting/shared-core'
import { renameSync } from 'fs'
import * as assert from 'power-assert'
import * as rmdir from 'rimraf'
import { finalize, tap } from 'rxjs/operators'

import { runCmd, RunCmdArgs } from '../src/index'


const rmdirAsync = promisify(rmdir)

const filename = basename(__filename)
const defaultCenterPath = normalize(join(userHome, '.myca'))
let defaultCenterPathBak = normalize(join(userHome, '.myca' + Math.random()))


const initArgs: RunCmdArgs = {
  cmd: 'init',
  options: null,
  debug: true,
}
const init$ = runCmd(initArgs)

describe(filename, () => {
  before(async () => {
    if (await isPathAccessible(defaultCenterPath)) {
      renameSync(defaultCenterPath, defaultCenterPathBak)
      console.info(
        '\n-----------\n',
        `Backup existing myca deafaultCenter ori path: "${defaultCenterPath}"\n`,
        `to back path: "${defaultCenterPathBak}"`,
        '\n-----------\n',
      )
    }
    else {
      defaultCenterPathBak = ''
    }
    // await createDirAsync(tmpDir)
  })
  after(async () => {
    try {
      await rmdirAsync(defaultCenterPath)
    }
    catch (ex) {
      console.error(ex)
      if (defaultCenterPathBak) {
        console.info(
          '\n----------\n',
          'Caution:\n',
          `Testing path can NOT be removed, path: "${defaultCenterPath}"\n`,
          `Default centerPat can NOT be resotre to original name, path: "${defaultCenterPathBak}"`,
          '\n----------\n',
        )
      }

    }
    if (defaultCenterPathBak && await isPathAccessible(defaultCenterPathBak)) {
      renameSync(defaultCenterPathBak, defaultCenterPath)
      console.info(
        '\n-----------\n',
        `Restore myca deafaultCenter from back path: "${defaultCenterPathBak}"\n`,
        `to ori path: "${defaultCenterPath}"`,
        '\n-----------\n',
      )
    }
  })


  describe('Should myca init works', () => {
    it('normal', done => {
      init$.pipe(
        tap((ret: string) => {
          assert(ret.includes('Default center created at path'), ret)
          assert(ret.includes(defaultCenterPath), ret)
        }),
        finalize(() => done()),
      ).subscribe()
    })
  })

  describe('Should myca initca works', () => {
    it('normal', done => {
      const args: RunCmdArgs = {
        cmd: 'initca',
        options: {
          days: 10950,
          pass: 'mycapass',
          CN: 'my root ca',
          O: 'my company',
          C: 'CN',

        },
        debug: false,
      }

      runCmd(args).pipe(
        tap((ret: string) => {
          assert(ret.includes('CA certificate created with:'), ret)
          assert(ret.includes('centerName: "default"'), ret)
          assert(ret.includes('crtFile'), ret)
          assert(ret.includes('privateKeyFile'), ret)
          assert(ret.includes(defaultCenterPath), ret)
        }),
        finalize(() => done()),
      ).subscribe()
    })
  })

})


test/20_helper.test.ts
/// <reference types="mocha" />

import {
  basename,
} from '@waiting/shared-core'
import * as assert from 'power-assert'
import { from as ofrom } from 'rxjs'
import { finalize, tap } from 'rxjs/operators'

import { cmdSet } from '../src/lib/config'
import { genCmdHelp } from '../src/lib/helper'


const filename = basename(__filename)

describe(filename, () => {
  describe('Should myca help works', () => {
    it('without args', done => {
      const help = genCmdHelp('')
      assert(help.includes('Standard commands'), help)

      return ofrom(cmdSet).pipe(
        tap(cmd => {
          assert(help.includes(cmd), help)
        }),
        finalize(() => done()),
      ).subscribe()
    })

    it('with initca -h', () => {
      const help = genCmdHelp('initca')
      assert(help.includes('--centerName'), help)
    })

    it('with issue -h', () => {
      const help = genCmdHelp('issue')
      assert(help.includes('--kind'), help)
    })

    it('with initcenter -h', () => {
      const help = genCmdHelp('initcenter')
      assert(help.includes('--path'), help)
    })

  })
})


src/lib/index.ts
import { log } from '@waiting/log'
import * as myca from 'myca'
import { Observable } from 'rxjs'
import { map, mergeMap } from 'rxjs/operators'

import { InitCenterArgs, RunCmdArgs } from './model'


export function runCmd(args: RunCmdArgs): Observable<string> {
  const { cmd, options, debug } = args

  debug && options && log(options)
  switch (cmd) {
    case 'init':
      return init()

    case 'initca':
      return initCa(<myca.CaOpts> options)

    case 'issue':
      return issue(<myca.CertOpts> options)

    case 'initcenter':
      return initCenter(<InitCenterArgs> options)

    default:
      throw new Error(`invalid cmd: "${cmd}"`)
  }
}


function init(): Observable<string> {
  return myca.initDefaultCenter().pipe(
    map(centerPath => `Default center created at path: "${centerPath}"`),
  )
}


function initCa(options: myca.CaOpts): Observable<string> {
  return myca.initCaCert(options).pipe(
    map(certRet => {
      return `CA certificate created with:
  centerName: "${certRet.centerName}"
  crtFile: "${certRet.crtFile}"
  privateKeyFile: "${certRet.privateKeyFile}"
    `
    }),
  )
}


function issue(options: myca.CertOpts): Observable<string> {
  // tslint:disable:max-line-length
  return myca.genCert(options).pipe(
    map(ret => {
      return `Issue a Certificate with:
  pubKey: \n${ret.pubKey}\n
  pass: "${ret.pass}" ${ options.kind === 'server' ? `\n  privateKeyFile: "${ret.privateKeyFile}"` : ''}
  privateKeyFile: "${ret.privateKeyFile}" ${ options.kind === 'server' ? `\n  privateUnsecureKeyFile: "${ret.privateUnsecureKeyFile}"` : ''}
  centerName: "${ret.centerName}"
  caKeyFile: "${ret.caKeyFile}"
  caCrtFile: "${ret.caCrtFile}"
  csrFile: "${ret.csrFile}"
  crtFile: "${ret.crtFile}"
  ${ options.kind === 'client' ? `pfxFile: "${ret.pfxFile}"` : ''}
    `
    }),
  )
}


function initCenter(options: InitCenterArgs): Observable<string> {
  const { name, path } = options

  return myca.initCenter(name, path).pipe(
    mergeMap(() => myca.getCenterPath(name)),
    map(centerPath => {
      return `center created with:
  centerName: "${name}"
  path: "${centerPath}"
    `
    }),
  )
}


src/lib/parse-opts.ts
import * as myca from 'myca'
import * as yargs from 'yargs'

import { cmdSet, initialCliArgs } from './config'
import { helpDefault } from './helper'
import { CliArgs, CmdType, InitCenterArgs } from './model'


export function parseCliArgs(argv: typeof yargs.argv): CliArgs {
  const args: CliArgs = { ...initialCliArgs }
  const cmdArr: string[] = argv._

  args.cmd = parseCmd(cmdArr)
  args.options = null
  args.needHelp = argv.h ? true : false
  args.debug = argv.d ? true : false

  return args
}


function parseCmd(args: string[]): CmdType {
  let command = ''

  for (let cmd of args) {
    cmd = cmd.toLowerCase()

    if (cmdSet.has(cmd)) {
      if (command) {
        throw new Error(`Duplicate command: "${cmd}" and "${command}"`)
      }
      else {
        command = cmd
      }
    }
  }

  if (! command) {
    const help = helpDefault()
    throw new Error(help)
  }
  return <CmdType> command
}


export function parseOpts(cmd: string, options: {[prop: string]: string | number}): CliArgs['options'] {
  if (cmd === 'init') {
    throw new Error('cmd should not be "init"')
  }

  if (cmd === 'initcenter') {
    return parseInitCenter(options)
  }

  const caOpts: CliArgs['options'] = cmd === 'initca'
    ? { ...myca.initialCaOpts }
    : { ...myca.initialCertOpts }
  const propMap = <Map<string, string>> new Map() // <upperKey, oriKey>

  Object.keys(caOpts).forEach(key => {
    propMap.set(key.toUpperCase(), key)
  })

  Object.keys(options).forEach(key => {
    const upperKey = key.toUpperCase()

    if (propMap.has(upperKey)) {
      Object.defineProperty(caOpts, <string> propMap.get(upperKey), {
        configurable: true,
        enumerable: true,
        writable: true,
        value: options[key],
      })
    }
  })

  if (cmd === 'issue') {
    if (typeof caOpts.SAN !== 'undefined') {
      const arr = parseMultiValue(caOpts.SAN)

      if (arr && arr.length) {
        caOpts.SAN = arr
      }
      else {
        delete caOpts.SAN
      }
    }
    if (typeof caOpts.ips !== 'undefined') {
      const arr = parseMultiValue(caOpts.ips)

      if (arr && arr.length) {
        caOpts.ips = arr
      }
      else {
        delete caOpts.ips
      }
    }
  }

  return caOpts
}


function parseMultiValue(arg: any): string[] {
  const arr = arg ? String(arg).split(',') : []
  const ret: string[] = []

  if (arr.length) {
    for (let value of arr) {
      value = value.trim()
      if (value) {
        ret.push(value)
      }
    }
  }

  return ret
}


function parseInitCenter(args: any): InitCenterArgs {
  const { path } = args
  let { name } = args

  name = String(name)
  if (! name) {
    throw new Error('value of name empty')
  }

  return { name, path }
}


src/lib/helper.ts
import { cmdSet, tw1, tw2, tw3 } from './config'
import { CmdType } from './model'


export function genCmdHelp(command: CmdType | ''): string {
  switch (command) {
    case 'init':
      return helpInit()

    case 'initca':
      return helpInitca()

    case 'initcenter':
      return helpInitcenter()

    case 'issue':
      return helpIssue()

    default:
      return helpDefault()
  }
}


export function helpDefault(): string {
  const head = 'Standard commands\n'
  const arr = Array.from(cmdSet)
  const more = 'More help: myca <command> -h'

  return `${head}${ arr.join('\t')}\n\n${more}`
}


export function helpInit(): string {
  const head = 'Usage: init [options]'
  const body = 'Valid options are:'
  const opts = [
    ` -h${tw2}Display this summary`,
    ` -d${tw2}Display debug info`,
  ]

  return `${head}\n${body}\n${ opts.join('\n') }`
}


export function helpInitca(): string {
  const head = 'Usage: initca [options]'
  const body = 'Valid options are:'
  const opts = [
    ` -h${tw3}Display this summary`,
    ` -d${tw3}Display debug info`,
    ` --centerName${tw2}Optinal. Contains no space. Default value default`,
    ` --alg${tw3}Optinal values: rsa | ec. Default value rsa`,
    ` --days${tw3}Default value 10950 (30years)`,
    ` --pass${tw3}Phrase greater then 4 chars`,
    ` --keyBits${tw2}Optional. Default value 4096 for alg rsa`,
    ` --ecParamgenCurve${tw1}Optional values: P-256 | P-384. Default value P-256 for alg ec`,
    ` --hash${tw3}Optional values: sha256 | sha384. Default value sha384\n`,

    ` --CN${tw3}Common Name. eg: --CN "Foo Bar"`,
    ` --OU${tw3}Optional. Organizational Unit Name (eg, section)`,
    ` --O${tw3}Optional. Organization Name (eg, company)`,
    ` --C${tw3}Optional. Country Name (2 letter code)`,
    ` --ST${tw3}Optional. State or Province Name (full name)`,
    ` --L${tw3}Optional. Locality Name (eg, city)`,
    ` --emailAddress${tw2}Optional`,
  ]

  return `${head}\n${body}\n${ opts.join('\n') }`
}


export function helpInitcenter(): string {
  const head = 'Usage: initcenter [options]'
  const body = 'Valid options are:'
  const opts = [
    ` -h${tw2}Display this summary`,
    ` -d${tw2}Display debug info`,
    ` --name${tw2}Center Name`,
    ` --path${tw2}Optional. Full path of the Center`,
  ]

  return `${head}\n${body}\n${ opts.join('\n') }`
}


export function helpIssue(): string {
  const head = 'Usage: issue [options]'
  const body = 'Valid options are:'
  const opts = [
    ` -h${tw3}Display this summary`,
    ` -d${tw3}Display debug info`,
    ` --kind${tw3}Values: 'ca' | 'server' | 'client'. Default value server`,
    ` --centerName${tw2}Optinal. Contains no space. Default value default`,
    ` --caKeyPass${tw2}Phrase for reading CA private key`,
    ` --alg${tw3}Optinal values: rsa | ec. Default value rsa`,
    ` --days${tw3}Default value 720 (2years)`,
    ` --pass${tw3}Phrase greater then 4 chars`,
    ` --keyBits${tw2}Default value 2048 for alg rsa`,
    ` --ecParamgenCurve${tw1}Optional values: P-256 | P-384. Default value P-256 for alg ec`,
    ` --hash${tw3}Optional values: sha256 | sha384. Default value sha256\n`,

    ` --CN${tw3}Common Name. eg: --CN "Foo Bar"`,
    ` --OU${tw3}Optional. Organizational Unit Name (eg, section)`,
    ` --O${tw3}Optional. Organization Name (eg, company)`,
    ` --C${tw3}Country Name (2 letter code)`,
    ` --ST${tw3}Optional. State or Province Name (full name)`,
    ` --L${tw3}Optional. Locality Name (eg, city)`,
    ` --emailAddress${tw2}Optional`,
    ` --SAN${tw3}Optional. Multiple values split by comma without space.\n` +
      `${tw3} eg: --SAN waitingsong.com,www.waitingsong.com\n` +
      `${tw3} or --SAN=waitingsong.com,www.waitingsong.com`,
    ` --ips${tw3}Optional.Multiple values split by comma without space.\n` +
      `${tw3} eg: --ips 127.0.0.1,192.168.0.1\n` +
      `${tw3} or: --ips=127.0.0.1,192.168.0.1`,
  ]

  return `${head}\n${body}\n${ opts.join('\n') }`
}
